### PR TITLE
feat & ref: implement derf and derfc and refactored unary_functions

### PR DESCRIPTION
**Pull request by parth121101**
Towards: #492

### CODE REVIEW

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 1]
```diff
@@ -80,6 +80,54 @@
...
+            "args": [("real",)],
+            "ret_type_arg_idx": 0
+        }
+    ],
+    "Derf": [
+        {
+            "args": [("real",)],
+            "ret_type_arg_idx": 0
+        }
+    ],
```
Comment (certik) [inline src/libasr/intrinsic_func_registry_util_gen.py:118]: Is this double precision erf function? If so, then it should be just `erf`.

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -42,7 +42,9 @@ enum class IntrinsicElementalFunctions : int64_t {
     Acosh,
     Atanh,
     Erf,
+    Derf,
     Erfc,
+    Derfc,
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:47]: These should not be added here, just use `Erf` and `Erfc`. Assuming it is the same issue as with `dsin` which just maps to `Sin`.

Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_functions.h:47]: Yes, I think we should just map to `erf` and `erfc` and reuse their implementation.

**Review by certik — APPROVED** [review]
I think this looks great to me, thanks!

File: integration_tests/intrinsics_228.f90 [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    y = -3.1473863781_dp
+    z = -7389.83936383_dp
+
+    print *, derf(x)
+    if (abs(derf(x) - (-1.00000000000000000e+00_dp)) > 1e-12) error stop
+    print *, derf(y)
+    if (abs(derf(y) - (-9.99991455910536065e-01_dp)) > 1e-12) error stop
+    print *, derf(z)
+    if (abs(derf(z) - (-1.00000000000000000e+00_dp)) > 1e-12) error stop
+
```
Comment (Shaikh-Ubaid) [inline integration_tests/intrinsics_228.f90:26]: Can we also check compile-time evaluation for `derf` and `derfc`? We can do that in a new PR.

**Comment by parth121101** [comment]
@Shaikh-Ubaid, @certik as suggested by you to just map derf and derfc with erf and erfc in the similar manner as that of dsin.
```
cat a.f90
program main
    print *, dsin([1.0_8, 2.0_8])
end program                 

                                                                                                                                       
gfortran a.f90 && ./a.out
  0.84147098480789650       0.90929742682568171     


lfortran a.f90           
semantic error: Argument 1 of dsin must be of double precision real type
 --> a.f90:2:14
  |
2 |     print *, dsin([1.0_8, 2.0_8])
  |              ^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.

```

This is what the result is when we are trying to add compile time tests in the form of an array. Do you think we need to make it an elemental function ?

**Comment by certik** [comment]
This should happen automatically. See why it would be different to just `sin([1.0_8, 2.0_8])`, the code should go via the same path.

**Pull request merged**